native/native_main.nim

```diff
@@ -0,0 +1,240 @@
+# compile with 
+# nim c --verbosity:0 native_main.nim
+
+# test with
+# time printf '%c\0\0\0{"cmd": "run", "command": "echo $PATH"}' 39 | ./native_main
+
+# compare with
+# time printf '%c\0\0\0{"cmd": "run", "command": "echo $PATH"}' 39 | python ~/.local/share/tridactyl/native_main.py
+
+import json
+import options
+import osproc
+import streams
+import os
+import posix
+# import endians
+import struct # nimble install struct
+
+const VERSION = "0.2.0"
+
+type 
+    MessageRecv* = object
+        cmd*, version*, content*, error*, command*, variable*, file*, dir*, to*, origin*: Option[string]
+        code: Option[int]
+type 
+    MessageResp* = object
+        cmd*, version*, content*, error*, command*: Option[string]
+        code: Option[int]
+
+# let a = MessageResp(cmd: some(""))
+# echo(a)
+
+proc trySwapJsonKey(json: JsonNode, old: string, nouveau: string) =
+    try:
+        json[nouveau] = json[old]
+        delete(json, old)
+    except KeyError:
+        discard
+
+
+proc getMessage(strm: Stream): MessageRecv =
+
+    try:
+        var length: int32
+        read(strm,length)
+        write(stderr, "Reading message length: " & $length & "\n")
+        if length == 0:
+            write(stderr, "No further messages, quitting.\n")
+            close(strm)
+            quit(0)
+
+        let message = readStr(strm, length)
+        write(stderr, "Got message: " & message & "\n")
+        var raw_json = parseJson(message)
+
+        # Compatibility with Python native messenger:
+        # rename env's _var_ key to _variable_ coz _var_ is reserved in Nim
+        trySwapJsonKey(raw_json, "var", "variable")
+        trySwapJsonKey(raw_json, "from", "origin")
+
+        return to(raw_json,MessageRecv)
+
+    except IOError:
+        write(stderr, "IO error - no further messages, quitting.\n")
+        close(strm)
+        quit(0)
+
+
+proc findUserConfigFile(): Option[string] =
+    let config_dir = getenv("XDG_CONFIG_HOME", expandTilde("~/.config"))
+    let candidate_files = [
+        config_dir / "tridactyl" / "tridactylrc",
+        getHomeDir() / ".tridactylrc",
+        getHomeDir() / "_config", "tridactyl" / "tridactylrc",
+        getHomeDir() / "_tridactylrc",
+    ]
+
+    var config_path = none(string)
+
+    for path in candidate_files:
+        if fileExists(path):
+            config_path = some(path)
+            break
+
+    return config_path
+
+
+proc handleMessage(msg: MessageRecv): string =
+
+    let cmd = msg.cmd.get()
+    var reply: MessageResp
+
+    case cmd:
+        of "version":
+            reply.version = some(VERSION)
+
+        of "getconfig":
+            write(stderr, "TODO: NOT IMPLEMENTED\n")
+
+        of "getconfigpath":
+            reply.content = findUserConfigFile()
+            reply.code = some(0)
+            if not isSome(reply.content):
+                reply.code = some(1)
+
+        of "run":
+            # this seems to use /bin/sh rather than the user's shell
```

Comment:
In the case of Windows, this appears to not use any normal shell at all, so you don't have access to any built-ins.

Reply:
Hmm, we should probably use `getenv "SHELL` on everything but Windows and hard-code `cmd.exe` there...